### PR TITLE
Aligments: explore idea on image block

### PR DESCRIPTION
**Pull request by ellatrix**
## Description

Explores ideas in #20650.

Screenshot 2020-03-05 at 14 35 56

(As you can see in the screenshot above, the block outline and toolbar are positioned correctly without an exception to the rule.)

The idea:

Add an "alignment wrapper" around the block instead of inside the block content. This alignment wrapper can similarly be used to position floated images relative to the main column.

Here's what the editor DOM looks like:

Screenshot 2020-03-05 at 14 44 00

As you can see it wraps around the block.

There are two ways to implement this:

* Either we automatically add this wrapper to both the `edit` and `save` output when we detect block support for it.
* Or the block add the wrapper by itself. A slight challenge here is that we automatically add `wp-block-*` classes to the top level element returned by `save`, which shouldn't be added to the wrapper.

Benefits:

* Much more straightforward to implement by blocks, especially if the wrapper would be added internally to both `edit` and `save` output.
* There's no exceptions needed to handle toolbar position and block selection styles.

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/block-library/src/image/edit.js [hunk 1]
```diff
@@ -430,29 +427,19 @@ export class ImageEdit extends Component {
...
-							mediaPlaceholder
-						) }
+						{ mediaPlaceholder }
 					</Block.div>
 				</>
 			);
 		}
 
-		const classes = classnames( className, {
+		const classes = classnames( className, `align${ align }`, {
```
Comment (youknowriad) [inline packages/block-library/src/image/edit.js]: I believe if we do this change we need to switch the classNames at the same time because of BC

**Review by youknowriad — COMMENTED** [review]
For the className automatically added to save, I've been wanting to remove the automatic addition for some time now to match edit and server. So might we can tie the removal of the magic addition with the lighterBlockWrapper opt-in behavior?

**Comment by ellatrix** [comment]
@youknowriad Yes, that sounds like a good idea. :) I was thinking it might be useful to have some sort of `Block.*` variant that can be used in the save function. This component could then add the class by default, and it future proofs anything we might want to add in the future.

**Comment by youknowriad** [comment]
Something unrelated I'm wondering, why did you go with `<Block.*` instead of `<Block tagName={"something"}` as it's a pattern that is widely used in our codebase?

**Comment by ellatrix** [comment]
So that it could be passed to e.g. `RichText`: `<RichText tagName={ Block.p } />`. This is a bit inspired by `react-spring`.

**Comment by ellatrix** [comment]
The could be seen as enhanced tags. :)

File: packages/block-library/src/image/edit.js (hunk 1 above)
Comment (ellatrix) [reply packages/block-library/src/image/edit.js]: I changed the classes.

File: packages/edit-post/src/style.scss [hunk 2]
```diff
@@ -104,13 +104,56 @@ body.block-editor-page {
...
 		max-width: none;
 	}
 }
+
+.wp-align-wrapper {
+	max-width: $content-width;
+	margin-left: auto;
+	margin-right: auto;
+
+	> .wp-block,
```
Comment (youknowriad) [inline packages/edit-post/src/style.scss]: why do we need this?

File: packages/edit-post/src/style.scss [hunk 3]
```diff
@@ -104,13 +104,56 @@ body.block-editor-page {
...
+	}
+
+	&.wp-align-left > *,
+	&.wp-align-right > * {
+		// Without z-index, won't be clickable as "above" adjacent content.
+		z-index: z-index(".block-editor-block-list__block {core/image aligned left or right}");
+	}
+
+	&.wp-align-center {
+		// Same as no alignment?
```
Comment (youknowriad) [inline packages/edit-post/src/style.scss]: this needs display: flex and justify-content center for me.

Comment (youknowriad) [reply packages/edit-post/src/style.scss]: you should be able to see the difference between "center" and "no" when you resize an image smaller than the $content-width

File: packages/block-library/src/image/edit.js [hunk 4]
```diff
@@ -516,17 +503,13 @@ export class ImageEdit extends Component {
...
-			? 'figure'
-			: Block.figure;
 		// Disable reason: Each block can be selected by clicking on it
 		/* eslint-disable jsx-a11y/click-events-have-key-events */
 		return (
 			<>
 				{ controls }
-				<AlignmentWrapper>
-					<BlockContentWrapper className={ classes }>
+				<div className={ `wp-align-wrapper wp-align-${ align }` }>
```
Comment (youknowriad) [inline packages/block-library/src/image/edit.js]: we're having `wp-align-undefined` sometimes

**Comment by youknowriad** [comment]
In my testing, floats are not working properly because it seems the width of the figure is always $content-size even if the image is small.

**Comment by youknowriad** [comment]
Seems like the breakage is because of these editor styles https://developer.wordpress.org/block-editor/developers/themes/theme-support/#changing-the-width-of-the-editor

We need to figure a new way for these?

**Comment by youknowriad** [comment]
One potential solution is to extract the styles defining the widths from here https://github.com/WordPress/gutenberg/pull/20656/files#diff-556b1933de5650dc555c469c584accc2R116 and move them to `editor-styles.css`

another solution would be to rely on CSS variables (but maybe it's too soon for this)

File: packages/edit-post/src/style.scss (hunk 2 above)
Comment (ellatrix) [reply packages/edit-post/src/style.scss]: Because the max-width should apply to the align wrapper, but not the nested block

**Comment by ellatrix** [comment]
> For the className automatically added to save, I've been wanting to remove the automatic addition for some time now to match edit and server. So might we can tie the removal of the magic addition with the lighterBlockWrapper opt-in behavior?

What's the behaviour for server? What do you envision for `save`? I was thinking of mirroring `Block.*` in the save function as well, which would automatically have the classes and filtered props etc. What do you think?

File: packages/edit-post/src/style.scss (hunk 3 above)
Comment (ellatrix) [reply packages/edit-post/src/style.scss]: display: flex has the unfortunate side effect of not collapsing margins. :) I'll figure it out though.

**Comment by youknowriad** [comment]
> I was thinking of mirroring Block.* in the save function as well

This would work, In the server we do nothing, we rely on the block author.

**Comment by youknowriad** [comment]
So I've been thinking about this more and we have two possibilities I think:

**First approach**

To have the following styles in editor styles. Instead of asking theme authors to add this to their editor styles, we could potentially have a CSS transform to generate it from https://developer.wordpress.org/block-editor/developers/themes/theme-support/#changing-the-width-of-the-editor

It's tricky though and there might be things we'll be missing.

```
.wp-align-wrapper {
	max-width: $content-width;
}

.wp-align-wrapper > .wp-block,
.wp-align-wrapper.wp-align-full {
	max-width: none;
}

.wp-align-wrapper.wp-align-wide {
	max-width: 1100px;
}
```

**Second approach**

Apply the `wp-block` class to the alignment wrapper in the editor. I think it makes sense if we consider that the main/only use-case for this class (why it exists) is to serve the purpose of assigning block widths in different alignment situations. Now, when we do this should we remove it from the block wrapper `Block.*` too (so it needs to know about the existence of the alignment support, I wonder if it could even be responsible for the alignment support: adding the wrapper...)

What's left is a CSS transform for editor styles to match `[data-align="full"]` `[data-align="wide"]` to the new classNames, unless we also keep these data attribute for BC.

I'm leaning towards the second approach with `Block.*` responsible for the alignment wrapper addition. What do you think?

**Comment by annezazu** [comment]
Closing this PR out as the original issue referenced has closed: 

https://github.com/WordPress/gutenberg/issues/20650#issuecomment-801743442

If that's incorrect, feel free to reopen!

**Pull request closed without merging**